### PR TITLE
ENH: Bounds around brain

### LINKED ISSUE #9755: ENH: Intentional boundaries of MRI and CT
Currently in the CT GUI code, the boundaries of the MRI/CT are visible only coincidentally (and for some data) because they show up only due to differences in 1) background / "zero" value, and/or 2) constant-value extrapolation when plotting. A more principled way to show the same information more clearly is:

1. allow setting the vmin/vmax of each image so that background noise can be effectively zeroed out / peak values of interest more visible
2. determine the bounding box of the given slice in each shown slice, and plot it around the slice

This will take a little bit of UI tweaking (vmin/vmax) and geometrical wrangling (putting in a matplotlib rectangle with the correct rotation, shape, origin, etc.) but I think it will be a nice aesthetic and usability boost.

Might be good to do after #9710

### PR DESCRIPTION
**Pull request by larsoner**
Closes #9755

When working on that, I found a bunch of small stuff that I thought should be fixed:

1. Add `cval` arg to `apply_volume_registration` so that you can choose what value to use/assume outside the bounds of the image domain during interpolation. (Naming chosen to match `scipy.ndimage` convention, but we could change it.) For the CT image the background value is -1024, and passing `cval='1%'` uses the first percentile of the image data (which is -1024).
2. Refactor iEEG GUI to be much more DRY by:

    1. Routing all `self._ras` changes through a `self._update_ras(ras)` that does clipping to valid (in-image) values only
    2. Using a `self._vox` property to avoid repeated `apply_trans(...).round().astype(int)`
    3. Making `self._current_slice` a property as well

    This should also help ensure that these variables remain fully in sync.
3. Change behavior of iEEG GUI crosshairs: they should always be fixed to the domain of the image and not change with zoom
4. Add bounding box to image domain
5. Fix bug where clicking outside the image would lead to an error (now it snaps to the closest in-image location, thanks `_update_ras`!)
6. Fix bug where the images were plotted with extent `(0, img.shape[dim])` instead of `(-0.5, img.shape[dim] - 0.5)`. This reflects the fact that the voxel coordinates are for the *center* of each voxel, so the first voxel should occupy the limits `(-0.5, 0.5)` and not `(0, 1)` in the image. If the last voxel is 255 (`img.shape[dim] == 256`), its data should occupy `(254.5, 255.5)`. Hence a domain of `(-0.5, 255.5)` is correct while `(0, 256)` is off by 0.5.

`main` after one click shows the bad cval effect as well as the weird/incorrect bounds of the crosshairs:

Screenshot from 2022-02-22 11-09-55

This PR after the same click shows a fixed cval and the crosshairs limited and fixed to the bounds of the domain:

Screenshot from 2022-02-22 11-57-30

Testing can be done with this little snippet:

<details>

```
# Minimal script to test the iEEG GUI

import os.path as op
import numpy as np
import nibabel as nib
import mne
misc_path = mne.datasets.misc.data_path(verbose=True)
raw = mne.io.read_raw(op.join(misc_path, 'seeg', 'sample_seeg_ieeg.fif'))
subj_trans = mne.coreg.estimate_head_mri_t(
    'sample_seeg', op.join(misc_path, 'seeg'))
reg_affine = np.array([
    [0.99270756, -0.03243313, 0.11610254, -133.094156],
    [0.04374389, 0.99439665, -0.09623816, -97.58320673],
    [-0.11233068, 0.10061512, 0.98856381, -84.45551601],
    [0., 0., 0., 1.]])
T1 = nib.load(op.join(misc_path, 'seeg', 'sample_seeg', 'mri', 'T1.mgz'))
CT_orig = nib.load(op.join(misc_path, 'seeg', 'sample_seeg_CT.mgz'))
CT_aligned = mne.transforms.apply_volume_registration(
    CT_orig, T1, reg_affine, cval='1%', verbose=True)
nib.save(CT_aligned, 'CT_aligned.nii')
CT_aligned = nib.load('CT_aligned.nii')
gui = mne.gui.locate_ieeg(raw.info, subj_trans, CT_aligned,
                          subject='sample_seeg',
                          subjects_dir=op.join(misc_path, 'seeg'))
```

</details>

Tutorial images also get updated

| main | PR |
| --- | --- |
|  | sphx_glr_10_ieeg_localize_004 |
|  | sphx_glr_10_ieeg_localize_005 |
|  | sphx_glr_10_ieeg_localize_006 |

### CODE REVIEW

**Comment by alexrockhill** [comment]
A lot of good fixes, my bad for putting some of those in there. Looks good but I'll look more in-depth when the CIs pass.

**Comment by alexrockhill** [comment]
Nice, just to be clear, the bounds issue doesn't invalidate any of the positions right? It's just plotting?

**Comment by larsoner** [comment]
> the bounds issue doesn't invalidate any of the positions right? It's just plotting?

It makes it so that if you click outside the CT, it's as if you clicked at the closest point inside the CT. I think this is the most sensible behavior...

**Comment by alexrockhill** [comment]
No that totally makes sense. I'm asking, there isn't anything that you've changed that would make positions found before the fixes not accurate as in off by half a voxel or something.

**Comment by alexrockhill** [comment]
Okay, when it merges, I will test it on the sample misc data and some other data and double check. I think it's not a big deal as the clicks are generally used to seed the `snap to center` feature but if that's turned off, it will change the positions slightly but I'm guessing not much.

**Comment by larsoner** [comment]
Feel free to review now -- the CircleCI failure is in `main` and I'm working to fix it in a separate forthcoming PR.

File: mne/gui/_ieeg_locate_gui.py [hunk 1]
```diff
@@ -307,43 +318,55 @@ def _save_ch_coords(self, info=None, verbose=None):
     def _plot_images(self):
         """Use the MRI and CT to make plots."""
         # Plot sagittal (0), coronal (1) or axial (2) view
-        self._images = dict(ct=list(), chs=list(),
-                            cursor=list(), cursor2=list())
+        self._images = dict(ct=list(), chs=list(), ct_bounds=list(),
```
Comment (alexrockhill) [inline mne/gui/_ieeg_locate_gui.py:321]: Great change! Not sure why this was so inaptly named...

File: mne/gui/_ieeg_locate_gui.py [hunk 2]
```diff
@@ -1013,31 +1047,27 @@ def _key_press_event(self, event):
...
-            xyz = apply_trans(self._ras_vox_t, self._ras)
-            if axis == 0:
-                xyz[[1, 2]] = pos
-            elif axis == 1:
-                xyz[[0, 2]] = pos
-            elif axis == 2:
-                xyz[[0, 1]] = pos
-            self._ras = apply_trans(self._vox_ras_t, xyz)
-            self._move_cursors_to_pos()
+        if event.inaxes is self._figs[axis].axes[0]:
```
Comment (alexrockhill) [inline mne/gui/_ieeg_locate_gui.py:1062]: Very nice, much more elegant, I didn't know about this one, good to know for the future

File: mne/transforms.py [hunk 3]
```diff
@@ -1759,6 +1760,10 @@ def apply_volume_registration(moving, static, reg_affine, sdr_morph=None,
     interpolation : str
         Interpolation to be used during the interpolation.
         Can be "linear" (default) or "nearest".
+    cval : float | str
+        The constant value to assume exists outside the bounds of the
+        ``moving`` image domain. Can be a string percentage like ``'1%%'``
```
Comment (alexrockhill) [inline mne/transforms.py:1765]: Is the `%` escaped so that it reads `1%`? Since Circle failed, I can't check, this is more of a note to self to check

Comment (larsoner) [reply mne/transforms.py:1765]: Yes because of the `verbose` / `fill_doc`

**Comment by alexrockhill** [comment]
I tested the GUI and it works great, I think you fixed some issues with the cursor plotting not showing up/updating correctly.

It does look like the locations changed for some channels by maybe half a voxel, so we'll want to redo registration. I can make a PR to `mne-misc-data` for that (`main` on the right)

Screen Shot 2022-02-22 at 11 06 13 AM

**Comment by alexrockhill** [comment]
This is what I get when I type the location of LENT1 (the first channel) into Freeview's TkRegRAS:
Screen Shot 2022-02-22 at 5 17 00 PM

Here is `main`:
Screen Shot 2022-02-22 at 5 18 16 PM

And here is `gui`:
Screen Shot 2022-02-22 at 5 18 47 PM

I think the difference in location is caused by the extent being different, I'll comment on the line after this comment. This followed how Liberty did it in the ECoG GUI. It looks like `main` matches `freeview` so I think that's what we should go with unless I'm missing something.

File: mne/gui/_ieeg_locate_gui.py [hunk 4]
```diff
@@ -214,9 +221,13 @@ def _load_image_data(self, ct):
         self._ras_vox_t = np.linalg.inv(self._vox_ras_t)
 
         self._voxel_sizes = np.array(self._mri_data.shape)
-        self._img_ranges = [[0, self._voxel_sizes[1], 0, self._voxel_sizes[2]],
-                            [0, self._voxel_sizes[0], 0, self._voxel_sizes[2]],
-                            [0, self._voxel_sizes[0], 0, self._voxel_sizes[1]]]
+        self._img_ranges = [
```
Comment (alexrockhill) [inline mne/gui/_ieeg_locate_gui.py:224]: This maybe should be reverted. I get where your coming from and it does make sense but this matches `freeview`. I might be wrong though...

**Comment by alexrockhill** [comment]
Wait, I'm looking at it again and maybe it's just that the colored circle is off center, the cursors might be in the same place for both `main` and `gui`

**Comment by alexrockhill** [comment]
Okay, just a small fix, I went ahead and just added it since I couldn't suggest a revision on that part of the code.

**Review by alexrockhill — APPROVED** [review]

**Comment by larsoner** [comment]
Hmm, I don't see why 76830541de3f4cdb28094a0304b50a581227fbab should be necessary. It suggests that something is wrong with the data axes / extent or something. I'll do a little bit of testing in FreeView tomorrow to see if I can make sense of it.

My plan is really just to load up `fsaverage` because its vox_ras_t is the same as its vox_mri_t, and it's only round numbers (-1, +1, 128, -128, etc.). So telling it you want voxel 0, 0, 0 for example and zoom in, it should be clear if it's in the very corner of the image or 0.5 mm from the sides of the shown image. Same for 255 (it's either going to be 1mm from the edge of the shown image, 0.5 mm from the sides, or right up against the edge, which would be the weirdest case!). It's entirely possible I'm wrong about it the voxel locations being centers and not one edge or the other. (It would be weird of 0 was at one edge and 255 at the other, though.)

**Comment by alexrockhill** [comment]
Hmmm okay, it looks off-center in the GUI without the commit and it fixes it, but we really should have a solid idea of what we're doing :)

I'll let you figure out what makes the most sense when you have a chance.

**Comment by larsoner** [comment]
So freeview makes the 256 voxels fit in the span [0, 255] (i.e., 255 mm). Here are some voxels I entered manually for `fsaverage`, along with what they showed on the screen:
```
freeview -v T1.mgz:grayscale=-50,256
```
| Voxel | `[128, 128, 128]` | `[0, 0, 0]` | `[255, 255, 255]` |
| -- | -- | -- | -- |
| FreeView | Screenshot from 2022-02-23 08-11-47 | Screenshot from 2022-02-23 08-12-07 | Screenshot from 2022-02-23 08-12-19 |

Since FreeView makes 256 voxels (each supposed to be 1mm in width) take up 255mm worth of space... I think they must have a bug...

I tried `fsleyes` and it uses a center-of-voxel representation. 128,128,128 is perfectly centered, and when navigating toward the edges, the transitions occur at pixel boundaries, as expected.

> Hmmm okay, it looks off-center in the GUI without the commit and it fixes it

When you say "the GUI" do you mean this GUI, or FreeView? Given the above, I'm hesitant to use FreeView's representation / visualization for determining if things are aligned to the sub-mm level or not.

**Comment by larsoner** [comment]
>> Hmmm okay, it looks off-center in the GUI without the commit and it fixes it

> When you say "the GUI" do you mean this GUI, or FreeView? Given the above, I'm hesitant to use FreeView's representation / visualization for determining if things are aligned to the sub-mm level or not.

I pushed a commit that should fix the math. There was a bug where the channel image extent was the same as the voxel extent -- but this is not correct if you actually want the *centers* of the voxels to line up. For example, the channel image 0, 0, 0 voxel center should be at the same location as the MRI/CT image 0, 0, 0 voxel center. If the MRI/CT is 256 on each side 1 mm isotropic, and the channel image is 1024 each side (0.25 mm isotropic), then it actually needs a *smaller* extent -- instead of going from -0.5 to 255.5, it needs to go from -0.125 to 255.125.

Make sense?

**Comment by larsoner** [comment]
Couldn't resist the temptation to add a voxel text box, too. Seems nice to keep that for consistency with FreeSurfer, too

**Comment by alexrockhill** [comment]
I a bit shocked that Freesurfer has a bug for something this low-level but you're investigation made sense...

Is it just me or, when you zoom in for the GUI, do the contact images looks slightly off-center? Half a voxel high in all three views but only for some contacts (I assume close to 0.5 in voxels so lots of rounding).

**Comment by alexrockhill** [comment]
The voxel box is nice! If you put in fractional voxels it rounds though and I think for Freeview you can put in fractional voxels and the voxels will round but the RAS will reflect the fractional voxel.

**Comment by alexrockhill** [comment]
Nevermind, I looked in `main` and they were equally off-center (i.e. slightly due to rounding) not sure why several in a row happened to be the same direction but I think it's good.

**Comment by larsoner** [comment]
> I think for Freeview you can put in fractional voxels and the voxels will round but the RAS will reflect the fractional voxel.

Yeah I wasn't sure about this one... there is no reason that we *need* to round. I'll push a quick commit for this.

**Comment by alexrockhill** [comment]
Thanks, I can think of a use case where you have the voxel location of the electrode target.

**Comment by larsoner** [comment]
All green (the Linux test passed but for some reason status didn't update) @alexrockhill feel free to merge if you're happy!

**Comment by alexrockhill** [comment]
Looks awesome, thanks a ton @larsoner!

**Pull request merged**